fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexPrefetchRangeKeyValueCursor.java

```diff
@@ -0,0 +1,84 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.KeySelector;
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.MappedKeyValue;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.record.cursors.CursorLimitManager;
+import com.apple.foundationdb.subspace.Subspace;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link KeyValueCursor} that scans an index using the IndexPrefetch operation.
+ * This subclass of {@link KeyValueCursor} uses a different scan operation ({@link com.apple.foundationdb.Transaction#getMappedRange})
+ * and therefore the actual returned types of the scanned range are different too - they need to be parsed into records
+ * rather than Index entries. The returned values are MappedKeyValues and contain all splits of
+ * the record for a particular index entry.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class IndexPrefetchRangeKeyValueCursor extends KeyValueCursor {
+    private IndexPrefetchRangeKeyValueCursor(@Nonnull final FDBRecordContext context,
+                                             @Nonnull final AsyncIterator<KeyValue> iterator,
+                                             int prefixLength,
+                                             @Nonnull final CursorLimitManager limitManager,
+                                             int valuesLimit) {
+
+        super(context, iterator, prefixLength, limitManager, valuesLimit);
+    }
+
+    /**
+     * A Builder for the cursor. Note that this builder actually extends the superclass' builder, and does not create
+     * a {@link IndexPrefetchRangeKeyValueCursor} but rather a {@link KeyValueCursor}, by virtue of the fact that it
+     * only extends teh {@link #scanRange} method. This is by design and the created cursor is a {@link KeyValueCursor}
+     * with no behavior differences from the default.
+     */
+    @API(API.Status.EXPERIMENTAL)
+    public static class Builder extends KeyValueCursor.Builder {
+        // The HopInfo that is used for the getRangeAndFlatMap call
+        private final byte[] hopInfo;
+
+        private Builder(@Nonnull Subspace indexSubspace, @Nonnull byte[] hopInfo) {
+            super(indexSubspace);
+            this.hopInfo = hopInfo;
+        }
+
+        public static Builder newBuilder(@Nonnull Subspace indexSubspace, @Nonnull byte[] hopInfo) {
+            return new Builder(indexSubspace, hopInfo);
+        }
```

Comment:
Our general convention would be to put the `newBuilder` method on the to-be-built class (so you call `Foo.newBuilder()` to get a new `Foo.Builder`), though it looks like we took a somewhat non-standard approach with `KeyValueCursor` and made people call `withSubspace`. Maybe it's fine, then, if this is like this, or if it is something like `.withSubspaceAndMapper`

Reply:
Right. Added to the same issue above for the refactoring of the cursor hierarchy.